### PR TITLE
[Artifact] Enforce the update of artifacts with conflicting keys

### PR DESCRIPTION
**Pull request by yaelgen**
Adding a fix for: https://jira.iguazeng.com/browse/ML-3604

While the body key is a non-empty string, we will not allow POSTing  that includes "key" with bodies that contain a different "key".

### CODE REVIEW

File: mlrun/api/crud/artifacts.py [hunk 1]
```diff
@@ -49,6 +49,13 @@ def store_artifact(
                 f"Artifact with conflicting project name - {data['project']} while request project : {project}."
                 f"key={key}, uid={uid}, data={data}"
             )
+        if data.get("metadata"):
+            data_key = data["metadata"].get("key")
```
Comment (alonmr) [inline mlrun/api/crud/artifacts.py:53]:
```suggestion
        data_key = data.get("metadata", {}).get("key")
```

**Review by alonmr — CHANGES_REQUESTED** [review]
please see `mlrun.api.db.sqldb.db.SQLDB._process_artifact_dict_to_store`
there is a comparison between key and db_key. We should move it there so that both verifications will be in the same place.
Also, don't forget about legacy artifacts which are deprecated but should still be supported until 1.5.0.

File: tests/api/api/test_artifacts.py [hunk 2]
```diff
@@ -119,6 +119,23 @@ def test_store_artifact_backwards_compatibility(db: Session, client: TestClient)
...
+        STORE_API_ARTIFACTS_PATH.format(project=PROJECT, uid=UID, key=KEY, tag=TAG),
+        data=artifact.to_json(),
+    )
+    assert resp.status_code == HTTPStatus.OK.value
+
+    resp = client.post(
+        STORE_API_ARTIFACTS_PATH.format(project=PROJECT, uid=UID, key="test", tag=TAG),
+        data=artifact.to_json(),
+    )
+    assert resp.status_code == HTTPStatus.BAD_REQUEST.value
```
Comment (alonmr) [inline tests/api/api/test_artifacts.py:136]: assert the error message as well

File: mlrun/api/db/sqldb/db.py [hunk 3]
```diff
@@ -543,6 +543,12 @@ def _process_artifact_dict_to_store(artifact, key, iter=None):
             )
         if not db_key:
             artifact["spec"]["db_key"] = key
+
+        data_key = artifact.get("metadata", {}).get("key")
+        if (key and data_key) and data_key != key:
+            raise mlrun.errors.MLRunInvalidArgumentError(
+                f"Artifact with conflicting key name - {data_key} while request key : {key}."
```
Comment (alonmr) [inline mlrun/api/db/sqldb/db.py:550]: I suggest having a similar error message for both errors (here and in line 542)
and please add the keys to the message similarly above (and with the legacy artifact)
```suggestion
                f"Conflict between requested key ({key}) and key in artifact metadata ({data_key})"
```

**Review by alonmr — COMMENTED** [review]
Looks good! A couple of minor comments and I'm missing a test for legacy artifacts.

File: mlrun/api/db/sqldb/db.py [hunk 4]
```diff
@@ -539,10 +539,16 @@ def _process_artifact_dict_to_store(artifact, key, iter=None):
         db_key = artifact["spec"].get("db_key")
         if db_key and db_key != key:
             raise mlrun.errors.MLRunInvalidArgumentError(
-                "Conflict between requested key and key in artifact body"
+                f"Conflict between requested key ({key}) and key in artifact metadata ({db_key})"
```
Comment (liranbg) [inline mlrun/api/db/sqldb/db.py:542]: line 542, 550, and 577 have the same exact failure message. why? can you refine the messages so backtracking of why/where it failed would be more clear to developer? that way, would be easier to understand from tests what have failed.

**Review by liranbg — CHANGES_REQUESTED** [review]

**Pull request closed without merging**